DatadogCore/Sources/Core/Storage/FilesOrchestrator.swift

```diff
@@ -41,14 +37,11 @@ internal class FilesOrchestrator: FilesOrchestratorType {
     /// Telemetry interface.
     let telemetry: Telemetry
 
-    #if canImport(UIKit)
-    /// Application state publisher for enriching telemetry.
-    let applicationStatePublisher: ApplicationStatePublisher = .init()
-    var subscription: ContextValueSubscription?
-    #endif
     /// Flag indicating if the application is in background state.
-    @ReadWriteLock
-    var inBackground = false
+    var inBackground: Bool {
+        return contextProvider.read().applicationStateHistory.currentSnapshot.state == .background
+    }
+    let contextProvider: DatadogContextProvider
```

Comment:
Instead of calling `contextProvider.read()`, would it be possible to pass the `context` from upstream routines (as method param)? This `.read()` doesn't come for free:

https://github.com/DataDog/dd-sdk-ios/blob/8be9c4c9c20443d53a1849ccc984c343e8901869/DatadogCore/Sources/Core/Context/DatadogContextProvider.swift#L92-L98

Effectively, we will be calling context queue with every batch closed. 

@maxep What's your take on this? Isn't there a risk of reaching deadlock lock on context queue?

Reply:
Good catch! That's a very expensive task for only telemetry and could indeed lead to a deadlock since we don't know on which thread we are.

Preferable to use the [`applicationState`](https://developer.apple.com/documentation/uikit/uiapplication/1623003-applicationstate) from the `UIApplication`.

Reply:
> Preferable to use the [applicationState](https://developer.apple.com/documentation/uikit/uiapplication/1623003-applicationstate) from the UIApplication.

Could be an option as well 👌. @maciejburda pls make sure it is safe to use from BG thread.

From different angle tho, the `FilesOrchestrator` implements 3 main routines:
```swift
func getWritableFile(...) throws -> WritableFile
func getReadableFile(...) -> ReadableFile?
func delete(readableFile: ReadableFile, ...)
```
The first one is called from Event Write Context API and the other two are triggered from Data Upload Worker. In both these callers we know the context:
* [EVC context](https://github.com/DataDog/dd-sdk-ios/blob/8be9c4c9c20443d53a1849ccc984c343e8901869/DatadogCore/Sources/Core/DatadogCore.swift#L314-L319)
* [Data Upload Worker context](https://github.com/DataDog/dd-sdk-ios/blob/8be9c4c9c20443d53a1849ccc984c343e8901869/DatadogCore/Sources/Core/Upload/DataUploadWorker.swift#L60)

We could pass the `context` value from these calls downstream to `FilesOrchestratorType`, this way avoiding extra cost or dependency in `FilesOrchestrator`. Keep in mind that `DatadogContext` is immutable and thread-safe. @maxep WDYT?

Reply:
Ah you are right, `UIApplication.applicationState` must be called from the main thread 😭  [see this](https://github.com/DataDog/dd-sdk-ios/blob/develop/DatadogCore/Sources/Core/DatadogCore.swift#L397L401) 
So yea, injecting the context from caller is safer!

Reply:
Refactored to top level context getting